### PR TITLE
"Path is not parsed correctly if '[' is present in the path"

### PR DESCRIPTION
**Pull request by krishnayalavarthi**
# PR Summary

## PR Context

This PR helps to fix this issue that I raised. Path is not parsed correctly if '[' is present in the path #12168

Issues fixed:
Having '[' character in directory doesn't work. 

Create a directory: e:\[test
Repro 1:
Test-Path 'e:\[test\'
Output: Error
Expected Output: True

Repro 2:
Test-Path "e:\`[test\"
Output: Error
Expected Output: True

## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/regex.cs [hunk 1]
```diff
@@ -293,24 +293,43 @@ public static bool ContainsWildcardCharacters(string pattern)
...
                 if (IsWildcardChar(pattern[index]))
                 {
+                    // If '[' or ']' is present alone, then it should not be treated as wildcard chars
+                    //
+                    // Check if both the '[' and the ']' brackets are present.
+                    // If present they should be in order, first '[' and then ']'
+                    // Otherwise, just treat ('[' or ']') as normal literal
+
+					// Check if we got '['
+                    if ( IsLeftBracket(pattern[index]) )
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/regex.cs]:
```suggestion
                    if (IsLeftBracket(pattern[index]))
```

File: src/System.Management.Automation/engine/regex.cs [hunk 2]
```diff
@@ -293,24 +293,43 @@ public static bool ContainsWildcardCharacters(string pattern)
...
+                    {
+						// got '['.
+                        leftBracket = true;
+                        continue;
+                    }
+					// got the ']' without the '['
+                    else if ( (leftBracket == false) && IsRightBracket(pattern[index]) )
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/regex.cs]:
```suggestion
                    else if (!leftBracket && IsRightBracket(pattern[index]))
```

File: src/System.Management.Automation/engine/regex.cs [hunk 3]
```diff
@@ -403,6 +422,16 @@ private static bool IsWildcardChar(char ch)
             return (ch == '*') || (ch == '?') || (ch == '[') || (ch == ']');
         }
 
+        private static bool IsLeftBracket(char ch)
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/regex.cs]: Is there a need for these methods, rather than just a simple equality test above?

**Review by rjmholt — COMMENTED** [review]
This PR should add tests for examples where behaviour has changed.

Also we should consider it with respect to the discussion and work done in https://github.com/PowerShell/PowerShell/pull/7399

File: src/System.Management.Automation/engine/regex.cs (hunk 3 above)
Comment (krishnayalavarthi) [reply src/System.Management.Automation/engine/regex.cs]: Sure! Removed those methods.

**Comment by kwkam** [comment]
> This PR should add tests for examples where behaviour has changed.
> 
> Also we should consider it with respect to the discussion and work done in #7399

Actually it is #7407 which is split from #5896, 2 years ago

**Comment by iSazonov** [comment]
Please continue discussion in #12168.

**Pull request closed without merging**